### PR TITLE
Fix gloas transition

### LINKED ISSUE #11314: Reject pre-Gloas slots in get_ptc
https://github.com/ethereum/consensus-specs/pull/5652/

### PR DESCRIPTION
**Pull request by tbenr**
fixes #11314

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

---

> [!NOTE]
> **Medium Risk**
> Changes fork-boundary validation for PTC lookups and gossip; incorrect bounds could reject valid attestations or accept invalid pre-fork traffic, but scope is narrow and well-tested.
> 
> **Overview**
> Fixes **Gloas fork transition** handling for payload timeliness committee (PTC) data and payload attestation gossip.
> 
> **`getPtc`** now fails fast when the beacon state's fork is already Gloas but the requested slot's epoch is **before the fork epoch**, because the cached previous-epoch PTC window is zeroed at the fork. The check does not apply when the state's fork version is still pre-Gloas (transition states).
> 
> **Payload attestation gossip** rejects messages whose slot is not Gloas-eligible via new **`Spec.isPayloadAttestationAvailableAtSlot`**, wired through **`MiscHelpersGloas.isPayloadAttestationAvailable`**.
> 
> Tests cover pre-fork PTC failures, valid post-fork lookups, non-Gloas fork edge cases, and gossip rejection before the fork.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 2622b33089b97ee2c2b2e51daea0e3197dd23963. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/BeaconStateAccessorsGloasTest.java [hunk 1]
```diff
@@ -206,6 +207,57 @@ public void getIndexedPayloadAttestation_returnsEmptyIndicesWhenNoBitsSet() {
     assertThat(indexed.attestingIndices()).isEmpty();
   }
 
+  @Test
+  public void getPtc_throwsForSlotBeforeGloasFork() {
+    final UInt64 forkEpoch = UInt64.valueOf(10);
+    final UInt64 forkSlot = spec.computeStartSlotAtEpoch(forkEpoch);
+    final BeaconState state = stateWithGloasFork(forkSlot, forkEpoch);
+    final UInt64 preForkSlot = forkSlot.minus(1);
```
Comment (rolfyone) [inline ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/BeaconStateAccessorsGloasTest.java]: could just use `decrement()` instead of `minus(1)`

File: ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/BeaconStateAccessorsGloasTest.java [hunk 2]
```diff
@@ -206,6 +207,57 @@ public void getIndexedPayloadAttestation_returnsEmptyIndicesWhenNoBitsSet() {
...
+            () ->
+                beaconStateAccessors.getIndexedPayloadAttestation(
+                    state, payloadAttestation(preForkSlot, 0)))
+        .isInstanceOf(IllegalArgumentException.class);
+  }
+
+  @Test
+  public void getPtc_allowsPreviousEpochSlotAtOrAfterGloasFork() {
+    final UInt64 forkEpoch = UInt64.valueOf(10);
+    final UInt64 stateSlot = spec.computeStartSlotAtEpoch(forkEpoch.plus(1));
```
Comment (rolfyone) [inline ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/BeaconStateAccessorsGloasTest.java]: probably through file really, could use decrement / increment functions in uint64

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/BeaconStateAccessorsGloas.java [hunk 3]
```diff
@@ -234,6 +235,15 @@ public SszUInt64Vector computePtc(final BeaconState state, final UInt64 slot) {
   @Override
   public IntList getPtc(final BeaconState state, final UInt64 slot) {
     final UInt64 epoch = miscHelpers.computeEpochAtSlot(slot);
+    final Fork fork = state.getFork();
+    // The previous-epoch PTC is all zeros at the Gloas fork, so pre-fork slots are not queryable
+    if (fork.getCurrentVersion().equals(config.getGloasForkVersion())) {
```
Comment (rolfyone) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/BeaconStateAccessorsGloas.java:240]: need to be greater than or equal for heze compliance? otherwise we'd have to override in heze...

Comment (tbenr) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/BeaconStateAccessorsGloas.java:240]: I believe the intent is to cover the PTC "activation" transition. So haze wont have this problem and the check can be skipped

Comment (StefanBratanov) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/BeaconStateAccessorsGloas.java:240]: Technically this would never happen with our code, since we use atSlot() and atEpoch() when we call .getPtc(), so we always get the empty list, but it's ok to safeguard just in case

Comment (StefanBratanov) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/BeaconStateAccessorsGloas.java:240]: Why not just, spec.atEpoch(epoch).getMilestone() in this case?

Comment (StefanBratanov) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/BeaconStateAccessorsGloas.java:240]: Oh ignore me it's inside BeaconStateAccessorsGloas

Comment (tbenr) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/gloas/helpers/BeaconStateAccessorsGloas.java:240]: ye this is to protect from a first gloas block containing PTC attestations from fulu essentially (it won't be produced by teku)

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/payloadattestation/PayloadAttestationMessageGossipValidator.java [hunk 4]
```diff
@@ -165,6 +166,16 @@ public SafeFuture<InternalValidationResult> validate(
                     state.getValidators().size());
               }
 
+              /*
+               * [REJECT] The payload attestation slot is at or after the Gloas fork
+               */
+              if (spec.atSlot(data.getSlot()).getMilestone().isLessThan(SpecMilestone.GLOAS)) {
```
Comment (mehdi-aouadi) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/payloadattestation/PayloadAttestationMessageGossipValidator.java]: Can't we have a fork-aware helper here? Maybe something like `isPayloadAttestationMessageAvailableAtSlot` same as `isExecutionPayloadEnvelopeAvailableAtSlot` in `MiscHelpersGloas`

**Review by StefanBratanov — DISMISSED** [review]
LGTM, can use Mehdi's suggestions, I will reapprove

**Comment by tbenr** [comment]
@StefanBratanov @mehdi-aouadi I applied the suggestion

**Review by StefanBratanov — APPROVED** [review]
